### PR TITLE
Direct and persistent interfaces to Xpress

### PR DESCRIPTION
**Pull request by bknueven**
## Summary/Motivation:
Xpress is a commonly used commercial MIP solver. This change adds direct and persistent interfaces to Xpress, leveraging Xpress's python interface. Both are largely based on their Gurobi counterparts. Note that FICO distributes Xpress with a community license (5000 variable/constraint limit) on [PyPI](https://pypi.org/project/xpress/).

## Changes proposed in this PR:
- New solvers 'xpress_direct' and 'xpress_persistent'
- Tests for xpress_persistent
- DONE: Modify Pyomo's test harness to run Xpress

## Limitations:
- Xpress has methods for solving general NLPs and MINLPs. These new interfaces are limited to problems with linear and quadratic constraints/objectives.

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

**Comment by blnicho** [comment]
I opened #1444 to track adding Xpress to our testing infrastructure

**Comment by blnicho** [comment]
@bknueven Xpress has been added to our GitHub Actions and Travis tests

**Comment by bknueven** [comment]
It looks like xpress 8.9 is causing the python 3.6 and 3.7 tests to segfault.

**Comment by bknueven** [comment]
I spent some time attempting to correct the segfault issue on Python 3.6/3.7. I'm able to reproduce it on my system on Python 3.7. I was ultimately unsuccessful -- though I narrowed it down to a test in pyomo-model-libraries and the xpress_persistent test, the behavior was non-deterministic on my system.

I also attempted to re-factor the Xpress direct/persistent interfaces to use as few Xpress python objects as possible, to no avail.

Given that the issue only comes up on exit, and the tests still pass, would an acceptable solution be to remove the xpress installation from python 3.6/3.7? I was hoping the new release of Xpress (8.9.2) would resolve this issue, but that's clearly not the case.

**Comment by jsiirola** [comment]
@bknueven: which test is the most likely to cause a segfault?  I can try and take a look at it and see if I can reproduce it here...

**Comment by bknueven** [comment]
@jsiirola here's the subset of tests that reliably reproduce it on my system:
```
nosetests -v --with-xunit --xunit-file=TEST-pyomo.xml "--eval-attr=nightly and (not fragile)" ../pyomo-model-libraries/cute/testCUTE_baseline.py pyomo/solvers/tests/checks/test_xpress_persistent.py 
```
Separately they do fine. I'm running macOS with python 3.7.7.

File: pyomo/solvers/plugins/solvers/xpress_persistent.py [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+        """
+        # see PR #366 for discussion about handling indexed
+        # objects and keeping compatibility with the
+        # pyomo.kernel objects
+        #if var.is_indexed():
+        #    for child_var in var.values():
+        #        self.compile_var(child_var)
+        #    return
+        if var not in self._pyomo_var_to_solver_var_map:
+            raise ValueError('The Var provided to compile_var needs to be added first: {0}'.format(var))
```
[the remark this answers is not in the document]
Comment (bknueven) [reply pyomo/solvers/plugins/solvers/xpress_persistent.py]: Of course, thanks!

File: pyomo/solvers/plugins/solvers/xpress_persistent.py [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+
+        Returns
+        -------
+        control value or dictionary of control values
+        """
+        return self._solver_model.getControl(*args)
+
+    def write(self, filename, flags=''):
+        """
+        Write the model to a file (e.g., and lp file).
```
[the remark this answers is not in the document]
Comment (bknueven) [reply pyomo/solvers/plugins/solvers/xpress_persistent.py]: Right...this one was copied out of gurobi_persisent :-)

File: pyomo/solvers/plugins/solvers/xpress_direct.py [hunk 3]
```diff
@@ -0,0 +1,786 @@
...
+            # exception above!
+            print("Import of xpress failed - xpress message=" + str(e) + "\n")
+            self._python_api_exists = False
+            
+        self._range_constraints = set()
+
+        # TODO: this isn't a limit of XPRESS, which implements an SLP
+        #       method for NLPs. But it is a limit of *this* interface
+        self._max_obj_degree = 2
+        self._max_constraint_degree = 2
```
[the remark this answers is not in the document]
Comment (bknueven) [reply pyomo/solvers/plugins/solvers/xpress_direct.py:89]: I think this one could be modified to support SLP. Xpress has support in its Python expression system for trigonometric functions, exponents and logarithms, and general polynomials.

In addition modifying the method which get information from generate_standard_repn, probably some of the solution fetching logic would need to be revisited as well.

That said, this implementation is largely based on its Gurobi and CPLEX counterparts, which don't have such functionality. I would definitely need help to make it work.

**Comment by jsiirola** [comment]
Hmmm. So far I have not been able to reproduce locally with Python 3.7.5 on RHEL6 / RHEL7 and xpress 8.9.0.  I will try building Python 3.7.7 and see if that makes a difference.

**Pull request merged**